File: src/extras-ktexteditor/libsessionconfiginterface.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");

/// https://api.kde.org/ktexteditor-sessionconfiginterface.html
pub const ktexteditor__sessionconfiginterface = struct {
    /// New constructs a new KTextEditor::SessionConfigInterface object.
    ///
    ///
    pub fn New() QtC.KTextEditor__SessionConfigInterface {
        return qtc.KTextEditor__SessionConfigInterface_new();
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#readSessionConfig)
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup ```
    pub fn ReadSessionConfig(self: ?*anyopaque, config: ?*anyopaque) void {
        qtc.KTextEditor__SessionConfigInterface_ReadSessionConfig(@ptrCast(self), @ptrCast(config));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#readSessionConfig)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, callback: *const fn (self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup) callconv(.c) void ```
    pub fn OnReadSessionConfig(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.KTextEditor__SessionConfigInterface_OnReadSessionConfig(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#readSessionConfig)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup ```
    pub fn QBaseReadSessionConfig(self: ?*anyopaque, config: ?*anyopaque) void {
        qtc.KTextEditor__SessionConfigInterface_QBaseReadSessionConfig(@ptrCast(self), @ptrCast(config));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#writeSessionConfig)
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup ```
    pub fn WriteSessionConfig(self: ?*anyopaque, config: ?*anyopaque) void {
        qtc.KTextEditor__SessionConfigInterface_WriteSessionConfig(@ptrCast(self), @ptrCast(config));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#writeSessionConfig)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, callback: *const fn (self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup) callconv(.c) void ```
    pub fn OnWriteSessionConfig(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.KTextEditor__SessionConfigInterface_OnWriteSessionConfig(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-sessionconfiginterface.html#writeSessionConfig)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface, config: QtC.KConfigGroup ```
    pub fn QBaseWriteSessionConfig(self: ?*anyopaque, config: ?*anyopaque) void {
        qtc.KTextEditor__SessionConfigInterface_QBaseWriteSessionConfig(@ptrCast(self), @ptrCast(config));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KTextEditor__SessionConfigInterface ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KTextEditor__SessionConfigInterface_Delete(@ptrCast(self));
    }
};

File: src_original/extras-ktexteditor/libsessionconfiginterface.cpp
#include <KConfigGroup>
#define WORKAROUND_INNER_CLASS_DEFINITION_KTextEditor__SessionConfigInterface
#include <sessionconfiginterface.h>
#include "libsessionconfiginterface.h"
#include "libsessionconfiginterface.hxx"

KTextEditor__SessionConfigInterface* KTextEditor__SessionConfigInterface_new() {
    return new VirtualKTextEditorSessionConfigInterface();
}

void KTextEditor__SessionConfigInterface_ReadSessionConfig(KTextEditor__SessionConfigInterface* self, const KConfigGroup* config) {
    auto* vktexteditor__sessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditor__sessionconfiginterface && vktexteditor__sessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditor__sessionconfiginterface->readSessionConfig(*config);
    } else {
        ((VirtualKTextEditorSessionConfigInterface*)self)->readSessionConfig(*config);
    }
}

void KTextEditor__SessionConfigInterface_WriteSessionConfig(KTextEditor__SessionConfigInterface* self, KConfigGroup* config) {
    auto* vktexteditor__sessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditor__sessionconfiginterface && vktexteditor__sessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditor__sessionconfiginterface->writeSessionConfig(*config);
    } else {
        ((VirtualKTextEditorSessionConfigInterface*)self)->writeSessionConfig(*config);
    }
}

// Base class handler implementation
void KTextEditor__SessionConfigInterface_QBaseReadSessionConfig(KTextEditor__SessionConfigInterface* self, const KConfigGroup* config) {
    auto* vktexteditorsessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditorsessionconfiginterface && vktexteditorsessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditorsessionconfiginterface->setKTextEditor__SessionConfigInterface_ReadSessionConfig_IsBase(true);
        vktexteditorsessionconfiginterface->readSessionConfig(*config);
    } else {
        ((VirtualKTextEditorSessionConfigInterface*)self)->readSessionConfig(*config);
    }
}

// Auxiliary method to allow providing re-implementation
void KTextEditor__SessionConfigInterface_OnReadSessionConfig(KTextEditor__SessionConfigInterface* self, intptr_t slot) {
    auto* vktexteditorsessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditorsessionconfiginterface && vktexteditorsessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditorsessionconfiginterface->setKTextEditor__SessionConfigInterface_ReadSessionConfig_Callback(reinterpret_cast<VirtualKTextEditorSessionConfigInterface::KTextEditor__SessionConfigInterface_ReadSessionConfig_Callback>(slot));
    }
}

// Base class handler implementation
void KTextEditor__SessionConfigInterface_QBaseWriteSessionConfig(KTextEditor__SessionConfigInterface* self, KConfigGroup* config) {
    auto* vktexteditorsessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditorsessionconfiginterface && vktexteditorsessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditorsessionconfiginterface->setKTextEditor__SessionConfigInterface_WriteSessionConfig_IsBase(true);
        vktexteditorsessionconfiginterface->writeSessionConfig(*config);
    } else {
        ((VirtualKTextEditorSessionConfigInterface*)self)->writeSessionConfig(*config);
    }
}

// Auxiliary method to allow providing re-implementation
void KTextEditor__SessionConfigInterface_OnWriteSessionConfig(KTextEditor__SessionConfigInterface* self, intptr_t slot) {
    auto* vktexteditorsessionconfiginterface = dynamic_cast<VirtualKTextEditorSessionConfigInterface*>(self);
    if (vktexteditorsessionconfiginterface && vktexteditorsessionconfiginterface->isVirtualKTextEditorSessionConfigInterface) {
        vktexteditorsessionconfiginterface->setKTextEditor__SessionConfigInterface_WriteSessionConfig_Callback(reinterpret_cast<VirtualKTextEditorSessionConfigInterface::KTextEditor__SessionConfigInterface_WriteSessionConfig_Callback>(slot));
    }
}

void KTextEditor__SessionConfigInterface_Delete(KTextEditor__SessionConfigInterface* self) {
    delete self;
}
